components/places/src/import/fennec_bookmarks.rs

```diff
@@ -0,0 +1,300 @@
+
+    // Clear the mirror now, since we're about to fill it with data from the fennec
+    // connection.
+    log::debug!("Clearing mirror to prepare for import");
+    conn.execute_batch(&WIPE_MIRROR)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating mirror with the bookmarks roots");
+    crate::bookmark_sync::create_synced_bookmark_roots(&conn)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Creating staging table");
+    conn.execute_batch(&CREATE_STAGING_TABLE)?;
+
+    log::debug!("Importing from Fennec to staging table");
+    conn.execute_batch(&POPULATE_STAGING)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating missing entries in moz_places");
+    conn.execute_batch(&FILL_MOZ_PLACES)?;
+    scope.err_if_interrupted()?;
+
+    log::debug!("Populating mirror");
+    conn.execute_batch(&POPULATE_MIRROR)?;
+    scope.err_if_interrupted()?;
+
+    // Ideally we could just do this right after `CREATE_AND_POPULATE_STAGING`,
+    // but we have constraints on the mirror structure that prevent this (and
+    // there's probably nothing bad that can happen in this case anyway). We
+    // could turn use `PRAGMA defer_foreign_keys = true`, but since we commit
+    // everything in one go, that seems harder to debug.
+    log::debug!("Populating mirror structure");
+    conn.execute_batch(&POPULATE_MIRROR_STRUCTURE)?;
+    scope.err_if_interrupted()?;
+
+    // log::debug!("Detaching Fennec database");
+    // drop(auto_detach);
+    // scope.err_if_interrupted()?;
+
+    let store = BookmarksStore::new(&conn, &scope);
+    let mut merger = Merger::new(&store, Default::default());
+    // We're already in a transaction.
+    merger.set_external_transaction(true);
+    log::debug!("Merging with local records");
+    merger.merge()?;
+    scope.err_if_interrupted()?;
+
+    // Update last modification time, sync status, etc
+    log::debug!("Fixing up bookmarks");
+    conn.execute_batch(&FIXUP_MOZ_BOOKMARKS)?;
+    scope.err_if_interrupted()?;
+    log::debug!("Cleaning up mirror...");
+    clear_mirror_on_drop.execute_now()?;
+    log::debug!("Committing...");
+    tx.commit()?;
+
+    // Note: update_frecencies manages its own transaction, which is fine,
+    // since nothing that bad will happen if it is aborted.
+    log::debug!("Updating frecencies");
+    store.update_frecencies()?;
+
+    log::info!("Successfully imported bookmarks!");
+
+    auto_detach.execute_now()?;
+
+    Ok(())
+}
+
+#[derive(Clone, Copy, PartialEq, PartialOrd, Hash, Debug, Eq, Ord)]
+#[repr(u8)]
+pub enum FennecBookmarkType {
+    // https://searchfox.org/mozilla-central/rev/ec806131cb7bcd1c26c254d25cd5ab8a61b2aeb6/mobile/android/base/java/org/mozilla/gecko/db/BrowserContract.java#291-295
+    Folder = 0,
+    Bookmark = 1,
+    Separator = 2,
+    // Not supported
+    // Livemark = 3,
+    // Query = 4,
+}
+
+lazy_static::lazy_static! {
+    static ref WIPE_MIRROR: &'static str =
```

Comment:
Nit: Can this go outside the `lazy_static`, like `POPULATE_MIRROR_STRUCTURE` below?